taichi/backends/vulkan/vulkan_device.cpp

```diff
@@ -1507,67 +1530,66 @@ void VulkanStream::submit(CommandList *cmdlist_) {
   }
   */
 
-  VkPipelineStageFlags stage_flag{VK_PIPELINE_STAGE_ALL_COMMANDS_BIT};
-
   VkSubmitInfo submit_info{};
   submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
   submit_info.commandBufferCount = 1;
   submit_info.pCommandBuffers = &buffer->buffer;
 
-  if (last_semaphore_) {
-    submit_info.waitSemaphoreCount = 1;
-    submit_info.pWaitSemaphores = &last_semaphore_->semaphore;
-    submit_info.pWaitDstStageMask = &stage_flag;
+  std::vector<VkSemaphore> vk_wait_semaphores;
+  std::vector<VkPipelineStageFlags> vk_wait_stages;
+
+  for (const StreamSemaphore &sema_ : wait_semaphores) {
+    auto sema = std::static_pointer_cast<VulkanStreamSemaphoreObject>(sema_);
+    vk_wait_semaphores.push_back(sema->vkapi_ref->semaphore);
+    vk_wait_stages.push_back(VK_PIPELINE_STAGE_ALL_COMMANDS_BIT);
+    buffer->refs.push_back(sema->vkapi_ref);
   }
 
+  submit_info.pWaitSemaphores = vk_wait_semaphores.data();
+  submit_info.waitSemaphoreCount = vk_wait_semaphores.size();
+  submit_info.pWaitDstStageMask = vk_wait_stages.data();
+
   auto semaphore = vkapi::create_semaphore(buffer->device, 0);
-  last_semaphore_ = semaphore;
   buffer->refs.push_back(semaphore);
 
   submit_info.signalSemaphoreCount = 1;
   submit_info.pSignalSemaphores = &semaphore->semaphore;
 
-  submitted_cmdbuffers_.push_back(buffer);
-
-  BAIL_ON_VK_BAD_RESULT(vkQueueSubmit(queue_, /*submitCount=*/1, &submit_info,
-                                      /*fence=*/VK_NULL_HANDLE),
-                        "failed to submit command buffer");
-}
-
-void VulkanStream::submit_synced(CommandList *cmdlist) {
-  vkapi::IVkCommandBuffer buffer =
-      static_cast<VulkanCommandList *>(cmdlist)->finalize();
+  auto fence = vkapi::create_fence(buffer->device, 0);
 
-  VkSubmitInfo submit_info{};
-  submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
-  submit_info.commandBufferCount = 1;
-  submit_info.pCommandBuffers = &buffer->buffer;
-
-  VkPipelineStageFlags stage_flag{VK_PIPELINE_STAGE_ALL_COMMANDS_BIT};
+  // Resource tracking, check previously submitted commands
+  // FIXME: Figure out why it doesn't work
```

Comment:
OOC, how exactly is it not working?

Reply:
It's crashing randomly after a while, we might have forgotten some references counting somewhere.